warehouse/query-core/src/main/java/datawave/query/jexl/nodes/QueryPropertyMarker.java

```diff
@@ -3,28 +3,80 @@
 import com.google.common.collect.ImmutableSet;
 import datawave.query.jexl.JexlNodeFactory;
 import datawave.query.jexl.visitors.QueryPropertyMarkerVisitor;
+import org.apache.commons.jexl2.parser.ASTAndNode;
 import org.apache.commons.jexl2.parser.ASTDelayedPredicate;
 import org.apache.commons.jexl2.parser.ASTEvaluationOnly;
 import org.apache.commons.jexl2.parser.ASTReference;
 import org.apache.commons.jexl2.parser.ASTReferenceExpression;
 import org.apache.commons.jexl2.parser.JexlNode;
-import org.apache.commons.jexl2.parser.Parser;
+import org.apache.commons.jexl2.parser.JexlNodes;
 import org.apache.commons.jexl2.parser.ParserTreeConstants;
 
+import java.lang.reflect.Constructor;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 import java.util.Set;
+import java.util.function.Function;
 
 /**
  * This is a node that can be put in place of an underlying reference node to place a property on an underlying query sub-tree (e.g. ExceededValueThreshold)
  */
 public abstract class QueryPropertyMarker extends ASTReference {
     
+    /**
+     * Return the {@link QueryPropertyMarker}'s designated label. This method must be overridden in implementations of {@link QueryPropertyMarker} and is used
+     * to identify instances of {@link QueryPropertyMarker} nodes in a query tree.
+     * 
+     * @return the marker's label
+     */
     public static String label() {
         throw new IllegalStateException("Label hasn't been configured in subclass.");
     }
     
+    /**
+     * Create and return a new query property marker with the given source node. If the source has a parent, the source will be replaced with the new marker
+     * node in the parent's children.
+     * 
+     * @param source
+     *            the source node
+     * @param constructor
+     *            the marker type constructor
+     * @param <MARKER>
+     *            the {@link QueryPropertyMarker} type
+     * @return a new marker instance
+     */
+    public static <MARKER extends QueryPropertyMarker> MARKER create(JexlNode source, Function<JexlNode,MARKER> constructor) {
+        JexlNode parent = source.jjtGetParent();
+        MARKER marker = constructor.apply(source);
+        if (parent != null) {
+            JexlNodes.replaceChild(parent, source, marker);
+        }
+        return marker;
+    }
```

Comment:
It's currently possible to apply the same marker multiple times to a node and this is a case we need to handle.  See https://github.com/NationalSecurityAgency/datawave/pull/1271 for example unit tests. I think we can add an `isSourceAlreadyMarked()` method here.

Reply:
I can do this, but if we want to return the original source node, the return type for the `QueryPropertyMarker.create()` methods will have to be `JexlNode` rather than a `MARKER` specific type, since I can anticipate problems if we try to cast the source to a `MARKER` type. So instead of 
```java
BoundedRange node = BoundedRange.create(source);
```
we would have
```java
JexlNode node = BoundedRange.create(source);
```
If you're okay with that, I can make this change.

Reply:
The more I think about it the more I dislike my original suggestion. I'll work the 'already marked' case in 1271 after this is merged.